File: app/src/main/java/org/metabrainz/mobile/presentation/features/login/LoginViewModel.java
package org.metabrainz.mobile.presentation.features.login;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import org.metabrainz.mobile.data.repository.LoginRepository;
import org.metabrainz.mobile.data.sources.api.entities.AccessToken;
import org.metabrainz.mobile.data.sources.api.entities.userdata.UserInfo;

public class LoginViewModel extends ViewModel {

    private final LoginRepository repository = LoginRepository.getRepository();
    private MutableLiveData<AccessToken> accessTokenLiveData;
    private MutableLiveData<UserInfo> userInfoLiveData;

    public LoginViewModel() {
    }

    public MutableLiveData<AccessToken> getAccessTokenLiveData() {
        if (accessTokenLiveData == null) accessTokenLiveData = repository.getAccessTokenLiveData();
        return accessTokenLiveData;
    }

    public MutableLiveData<UserInfo> getUserInfoLiveData() {
        if (userInfoLiveData == null) userInfoLiveData = repository.getUserInfoLiveData();
        return userInfoLiveData;
    }

    public void fetchAccessToken(String code) {
        repository.fetchAccessToken(code);
    }

    public void fetchUserInfo() {
        repository.fetchUserInfo();
    }
}

File: app/src/main/java/org/metabrainz/mobile/presentation/features/login/LoginViewModel.kt
package org.metabrainz.mobile.presentation.features.login

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import org.metabrainz.mobile.data.repository.LoginRepository
import org.metabrainz.mobile.data.sources.api.entities.AccessToken
import org.metabrainz.mobile.data.sources.api.entities.userdata.UserInfo
import javax.inject.Inject

@HiltViewModel
class LoginViewModel @Inject constructor(private var repository: LoginRepository?) : ViewModel() {
    var accessTokenLiveData: MutableLiveData<AccessToken?>? = null
        get() {
            if (field == null) field = repository!!.accessTokenLiveData
            return field
        }
        private set
    var userInfoLiveData: MutableLiveData<UserInfo?>? = null
        get() {
            if (field == null) field = repository!!.userInfoLiveData
            return field
        }
        private set

    fun fetchAccessToken(code: String?) {
        repository!!.fetchAccessToken(code)
    }

    fun fetchUserInfo() {
        repository!!.fetchUserInfo()
    }

    override fun onCleared() {
        super.onCleared()
        repository = null
    }
}